### PR TITLE
tryDial timeout

### PR DESCRIPTION
**Pull request by vpikulik**

### CODE REVIEW

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: fasthttpproxy/http.go [hunk 1]
```diff
@@ -4,10 +4,10 @@ import (
 	"bufio"
 	"encoding/base64"
 	"fmt"
+	"github.com/valyala/fasthttp"
```
[the remark this answers is not in the document]
Comment (vpikulik) [reply fasthttpproxy/http.go]: done

File: tcpdialer.go [hunk 2]
```diff
@@ -333,7 +333,10 @@ func (d *TCPDialer) tryDial(network string, addr *net.TCPAddr, deadline time.Tim
 	ch := chv.(chan dialResult)
 	go func() {
 		var dr dialResult
-		dr.conn, dr.err = net.DialTCP(network, d.LocalAddr, addr)
+		dialer := net.Dialer{LocalAddr: d.LocalAddr}
+		ctx, cancel_ctx := context.WithDeadline(context.Background(), deadline)
+		defer cancel_ctx()
+		dr.conn, dr.err = dialer.DialContext(ctx, network, addr.String())
```
Comment (erikdubbelboer) [inline tcpdialer.go:339]: Now that a `context.WithDeadline` the use of the timer and goroutine isn't needed anymore. Can you please simplify the function by removing the goroutine and the bottom timer code. The timer on top waiting for `concurrencyCh` is still needed.

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: tcpdialer.go (hunk 2 above)
Comment (vpikulik) [reply tcpdialer.go]: please check the last commit

File: tcpdialer.go [hunk 3]
```diff
@@ -326,46 +326,16 @@ func (d *TCPDialer) tryDial(network string, addr *net.TCPAddr, deadline time.Tim
...
-		chv = make(chan dialResult, 1)
-	}
-	ch := chv.(chan dialResult)
-	go func() {
-		var dr dialResult
-		dr.conn, dr.err = net.DialTCP(network, d.LocalAddr, addr)
-		ch <- dr
-		if concurrencyCh != nil {
-			<-concurrencyCh
-		}
```
Comment (erikdubbelboer) [inline tcpdialer.go:340]: This is still needed after the `DialContext` returns (before the return out of the function).

**Review by erikdubbelboer — CHANGES_REQUESTED** [review]

File: tcpdialer.go (hunk 3 above)
Comment (vpikulik) [reply tcpdialer.go:340]: Done

**Comment by erikdubbelboer** [comment]
Thanks1

**Pull request merged**